[CHORE] Improve cooking building UI/UX

Polysemouse opened the pull request:
# Description

- for firepit, kitchen, bakery and deli:
  - standardize pixel size and fix pixel alignments
  - added progresss bar when crafting
  - on PC, popup is shown on hover to show remaining time
  - added ready indicator (!) when cooking is ready
- fix crafting machine not returning non-integer remaining crafting time and not executing right at the start of crafting
  - or else progress bar progress will be shifted by 1 pixels
- move arcade and quest to <Water /> component
  - external islands belongs to this component
- fix refreshing issues for UIRefresher

https://user-images.githubusercontent.com/107602352/203973534-fede9414-7f99-4abb-bedf-64d89d8035ba.mp4

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- complete a cooking cycle for the firepit, kitchen, bakery and deli respectively

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

hujikol commented:
Got struggling to understand and to consume the timer from crafting machine, got a lot of insight with this PR 🚀

Polysemouse commented on src/features/island/buildings/components/building/bakery/Bakery.tsx:
```diff
@@ -33,8 +35,11 @@ export const Bakery: React.FC<Props> = ({
-  if (!craftingService || !handleShowCraftingTimer)
+  useUiRefresher({ active: crafting });
+
+  if (!craftingService?.initialized || !handleShowCraftingTimer) {
```
here

spencerdezartsmith commented on src/features/island/buildings/lib/craftingMachine.ts:
```diff
@@ -144,7 +145,7 @@ export const craftingMachine = createMachine<
           const now = Date.now();
 
-          return Math.floor((readyAt - now) / 1000);
+          return (readyAt - now) / 1000;
```
This can end up being a floating point number which is why we were flooring it here. Is there a reason you have removed this?

Polysemouse replied:
The removal of Math.floor() is intentional:
  - It aligns with other components (plots, resource nodes remianing time) where the time left is haldled with a floating point.  The secondsToString() method is going to round the value down anyways in the UI
  - It solve the issue where the progress bar progress is displaying the wrong progress for several amounts of seconds
  - It solve the issue where the "Is it ready yet?" (or something that will replace this modal) modal isn't displaying for the last second in the progress

spencerdezartsmith commented on src/features/island/buildings/components/building/BuildingImageWrapper.tsx:
```diff
@@ -1,25 +1,152 @@
+  const [isMobile] = useIsMobile();
+
+  const onMouseEnter = () => {
+    true && setShowHover(true);
```
Is this statement intentional?

Polysemouse replied:
The true && can be removed.

The hover for building is reomved, so this component won't be in the PR anymore

spencerdezartsmith commented on src/features/island/buildings/components/building/BuildingImageWrapper.tsx:
```diff
@@ -1,25 +1,152 @@
+
+  // time to complete popup
+  // only shows on PC on hover
+  const TimeToComplete = () => {
```
We aren't massive fans of the hover information panels. The ux of these on mobile is not good and they can become annoying when you have a very crowded land where these are popping up a lot. I think we would prefer to not add another one in here. 

<img width="355" alt="Screen Shot 2022-11-28 at 12 25 07 pm" src="https://user-images.githubusercontent.com/17863697/204173625-6857f1a4-433e-4523-9814-7cd8073f3d69.png">

Polysemouse replied:
This has been removed in the recent commits

spencerdezartsmith commented on src/features/island/buildings/components/building/bakery/Bakery.tsx:
```diff
@@ -33,8 +35,11 @@ export const Bakery: React.FC<Props> = ({
   const { setToast } = useContext(ToastContext);
 
-  if (!craftingService || !handleShowCraftingTimer)
+  useUiRefresher({ active: crafting });
```
Why do we need this ui refresher here? We have actually built these components to NOT refresh on every tick but to instead only rerender when a state changes. We have done this by interpreting the machine and only subscribing to certain changes. As the machine has a tick implemented we could easily just read the machine and get the rerender out of the box.

Why is there a need for it?

Polysemouse replied:
The `useUiRefresher` was implemented because the progress bar won't update without the refresher.

I have updated the PR to consume the `secondsTillReady` properly for the buildings, so I have removed the `useUiRefresher` refrences from these buildings (see my latest commit)

spencerdezartsmith reviewed (changes requested):
Thanks for the PR. I don't think implementing a hover panel is the direction we want to go for UX improvements on these modals. I think that the progress bar is a good addition and then adding the time inside the kitchen modals as suggested in coders chat. 

I do think that the progress bar could move further up so that its on top of the building image. These progress bars can make the space super cluttered so having them as part of the image makes more sense.

I am also unsure about adding the `uiRefresher` into these buildings. I need to understand the context around this change? We had intentionally built these to not do this.

Polysemouse commented:
@spencerdezartsmith I have addressed the comments in the latest commits:
- The hover panel has been removed
- The progress bar has been moved up so it is contained entirely in the building squares
- The uiRefresher is removed for the 4 types of cooking buildings and the secondsTillReady property of the machine is used to update the progress bar instead

spencerdezartsmith commented:
Thanks for that. We will actually need to remove this code. We now have a few unused props in the `BuildingImageWrapper` one of which causes a re render every second something is cook on a building. By adding a useSelector on the `secondsTillReady` context variable is will cause each component wrapped with a crafting machine and all children to re render on every `TICK`. We don't want this behaviour which is why we used the `useInterpret` hook on this machine so we could avoid this. 

If you want to keep this work for a future revisit then maybe you could create a new branch and cherry pick the other changes. I think the ! is great and works well with the new in progress flow.

I am also getting a strange situation with this:

```
  if (!craftingService?.initialized) {
    return <img src={firePit} className="w-full" />;
  }
``` 

It's causing the page to load with no npc's for a few seconds. I think we can stick with returning `null` is `!craftingService`? Thoughts?
<img width="623" alt="Screen Shot 2022-11-29 at 10 44 46 am" src="https://user-images.githubusercontent.com/17863697/204404099-b34cfd80-34fb-4c7b-b39e-4b9cd4412f82.png">

Polysemouse commented:
I was using `!craftingService?.initialized` to avoid warnings in the console, but it seems like it's not needed after several local testing.  I have updated the PR according to your suggestions.

spencerdezartsmith commented on src/features/island/buildings/components/building/WithCraftingMachine.tsx:
```diff
@@ -15,12 +15,15 @@ const isIdle = (state: MachineState) => state.matches("idle");
 const isCrafting = (state: MachineState) => state.matches("crafting");
 const isReady = (state: MachineState) => state.matches("ready");
 const itemName = (state: MachineState) => state.context.name;
+const secondsTillReady = (state: MachineState) =>
```
We will need to remove this as well as the `const secondsLeft = useSelector(craftingService, secondsTillReady);` below as this is what subscribes us to this value.

Polysemouse replied:
Removed

spencerdezartsmith commented on src/features/island/buildings/components/building/WithCraftingMachine.tsx:
```diff
@@ -55,13 +58,15 @@ export const WithCraftingMachine = ({
     crafting,
     ready,
     name,
+    secondsLeft,
```
And this..

Polysemouse replied:
Removed

Polysemouse commented:
I have removed the extra code as suggested

spencerdezartsmith reviewed (approved):
Thanks for sticking this one out 🤗
